<a href="https://colab.research.google.com/github/fralfaro/MAT281/blob/main/docs/labs/lab_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# MAT281 - Laboratorio N°03





**Objetivo**: Aplicar técnicas avanzadas de manipulación y análisis de datos con pandas sobre un conjunto real de datos de contenido de Netflix, reforzando buenas prácticas y métodos eficientes sin recurrir a `groupby`, `merge`, `pivot`, ni `join`.



**Dataset**:

Trabajaremos con el archivo `netflix_titles.csv`, que contiene información sobre los títulos disponibles en la plataforma Netflix hasta el año 2021.

| Variable       | Clase     | Descripción                                                                 |
|----------------|-----------|------------------------------------------------------------------------------|
| show_id        | caracter  | Identificador único del título en el catálogo de Netflix.                   |
| type           | caracter  | Tipo de contenido: 'Movie' o 'TV Show'.                                     |
| title          | caracter  | Título del contenido.                                                       |
| director       | caracter  | Nombre del director (puede ser nulo).                                       |
| cast           | caracter  | Lista de actores principales (puede ser nulo).                              |
| country        | caracter  | País o países donde se produjo el contenido.                                |
| date_added     | fecha     | Fecha en la que el título fue agregado al catálogo de Netflix.              |
| release_year   | entero    | Año de lanzamiento original del título.                                     |
| rating         | caracter  | Clasificación por edad (por ejemplo: 'PG-13', 'TV-MA').                      |
| duration       | caracter  | Duración del contenido (minutos o número de temporadas para series).        |
| listed_in      | caracter  | Categorías o géneros en los que está clasificado el contenido.              |
| description    | caracter  | Breve sinopsis del contenido.                                               |




In [1]:
import pandas as pd

# Cargar datos
df = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/netflix_titles.csv')
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...



### Parte 1: Limpieza y preparación

1. Revisar y describir el dataset:

   * ¿Cuántas filas y columnas tiene?
   * ¿Qué tipos de datos hay?
   * ¿Cuántos valores nulos hay por columna?

2. Transformar la columna `date_added` a tipo fecha.

3. Crear columnas auxiliares con `assign`:

   * Año (`year_added`)
   * Mes (`month_added`)



In [2]:
# Parte 1: Limpieza y preparación

# 1. Revisar dimensiones del dataset
print("Cantidad de filas y columnas:")
print(df.shape)

print("\nTipos de datos por columna:")
display(df.dtypes)

print("\nValores nulos por columna:")
display(df.isnull().sum())

# 2. Transformar date_added a tipo fecha
df["date_added"] = pd.to_datetime(df["date_added"].str.strip(), errors="coerce")

print("\nTipo de dato de date_added después de la transformación:")
print(df["date_added"].dtype)

# 3. Crear columnas auxiliares con assign
df = df.assign(
    year_added = df["date_added"].dt.year,
    month_added = df["date_added"].dt.month
)

# Vista previa del dataset preparado
df.head()

Cantidad de filas y columnas:
(8807, 12)

Tipos de datos por columna:


,0
show_id,object
type,object
title,object
director,object
cast,object
country,object
date_added,object
release_year,int64
rating,object
duration,object



Valores nulos por columna:


,0
show_id,0
type,0
title,0
director,2634
cast,825
country,831
date_added,10
release_year,0
rating,4
duration,3



Tipo de dato de date_added después de la transformación:
datetime64[ns]


,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added,month_added
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021.0,9.0
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",2021.0,9.0
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,2021.0,9.0
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,2021-09-24,2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo...",2021.0,9.0
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,2021.0,9.0


En esta primera etapa se revisó la estructura general del dataset de Netflix, identificando su cantidad de filas, columnas, tipos de datos y valores faltantes. Este paso es importante porque permite conocer la calidad inicial de la información antes de aplicar transformaciones.

Luego, la columna `date_added` fue convertida a formato de fecha, lo que permite trabajar con componentes temporales como año y mes. A partir de esta columna se crearon `year_added` y `month_added`, variables auxiliares útiles para analizar cuándo se agregaron los títulos al catálogo.

## Parte 2: Técnicas avanzadas de pandas

4. Utilizar `.loc` para seleccionar películas (`type == 'Movie'`) que fueron agregadas después del año 2018.

5. Utilizar `str.contains()` y `str.extract()`:

   * Filtrar títulos que contienen la palabra 'love' (sin distinguir mayúsculas/minúsculas).
   * Extraer la duración en minutos para las películas desde la columna `duration`.

6. Aplicar `explode()` sobre la columna `listed_in` para obtener una fila por cada género.

7. Obtener un top 10 de géneros más frecuentes utilizando `value_counts()`.

8. Aplicar `where()` y `mask()` para marcar las películas de más de 120 minutos como contenido largo en una nueva columna.

9. Utilizar `.loc` para filtrar películas que cumplen con:

   * Más de 100 minutos de duración.
   * Rating igual a `'R'`.
   * País igual a `'United States'`.

10. Utilizar `.style` para formatear visualmente el top 10 de películas más largas.

In [3]:
# Parte 2: Técnicas avanzadas de pandas

# 4. Seleccionar películas agregadas después del año 2018 usando .loc
movies_after_2018 = df.loc[
    (df["type"] == "Movie") & (df["year_added"] > 2018)
]

print("Películas agregadas después de 2018:")
display(movies_after_2018[["title", "type", "date_added", "year_added"]].head())

# 5. Filtrar títulos que contienen la palabra 'love'
love_titles = df.loc[
    df["title"].str.contains("love", case=False, na=False)
]

print("\nTítulos que contienen la palabra 'love':")
display(love_titles[["title", "type", "release_year"]].head())

# Extraer duración en minutos para películas desde duration
df = df.assign(
    duration_minutes = df["duration"].str.extract(r"(\d+)").astype(float)
)

print("\nDuración extraída en minutos o temporadas:")
display(df[["title", "type", "duration", "duration_minutes"]].head())

# Para análisis de duración de películas, se filtra solo Movie
movies = df.loc[df["type"] == "Movie"].copy()

# 6. Aplicar explode sobre listed_in para obtener una fila por género
df_genres = df.assign(
    genre = df["listed_in"].str.split(", ")
).explode("genre")

print("\nDataset con una fila por género:")
display(df_genres[["title", "type", "genre", "rating"]].head(10))

# 7. Top 10 de géneros más frecuentes
top_10_genres = df_genres["genre"].value_counts().head(10)

print("\nTop 10 géneros más frecuentes:")
display(top_10_genres)

# 8. Usar where y mask para marcar películas de más de 120 minutos como contenido largo
movies = movies.assign(
    content_length = "Contenido normal"
)

movies["content_length"] = movies["content_length"].where(
    movies["duration_minutes"] <= 120,
    "Contenido largo"
)

# Alternativa con mask
movies["content_length_mask"] = "Contenido normal"
movies["content_length_mask"] = movies["content_length_mask"].mask(
    movies["duration_minutes"] > 120,
    "Contenido largo"
)

print("\nClasificación de duración de películas:")
display(movies[["title", "duration", "duration_minutes", "content_length", "content_length_mask"]].head())

# 9. Filtrar películas con más de 100 minutos, rating R y país United States
filtered_movies = movies.loc[
    (movies["duration_minutes"] > 100) &
    (movies["rating"] == "R") &
    (movies["country"].str.contains("United States", case=False, na=False))
]

print("\nPelículas con más de 100 minutos, rating R y país United States:")
display(filtered_movies[["title", "country", "rating", "duration_minutes"]].head(10))

# 10. Formatear visualmente el top 10 de películas más largas
top_10_longest_movies = movies.sort_values("duration_minutes", ascending=False).head(10)

top_10_longest_movies[["title", "country", "release_year", "rating", "duration_minutes"]].style.background_gradient(
    subset=["duration_minutes"],
    cmap="Blues"
).format({
    "duration_minutes": "{:.0f} min"
})

Películas agregadas después de 2018:


,title,type,date_added,year_added
0,Dick Johnson Is Dead,Movie,2021-09-25,2021.0
6,My Little Pony: A New Generation,Movie,2021-09-24,2021.0
7,Sankofa,Movie,2021-09-24,2021.0
9,The Starling,Movie,2021-09-24,2021.0
12,Je Suis Karl,Movie,2021-09-23,2021.0



Títulos que contienen la palabra 'love':


,title,type,release_year
25,Love on the Spectrum,TV Show,2021
158,Love Don't Cost a Thing,Movie,2003
159,Love in a Puff,Movie,2010
206,"LSD: Love, Sex Aur Dhokha",Movie,2010
227,Really Love,Movie,2020



Duración extraída en minutos o temporadas:


,title,type,duration,duration_minutes
0,Dick Johnson Is Dead,Movie,90 min,90.0
1,Blood & Water,TV Show,2 Seasons,2.0
2,Ganglands,TV Show,1 Season,1.0
3,Jailbirds New Orleans,TV Show,1 Season,1.0
4,Kota Factory,TV Show,2 Seasons,2.0



Dataset con una fila por género:


,title,type,genre,rating
0,Dick Johnson Is Dead,Movie,Documentaries,PG-13
1,Blood & Water,TV Show,International TV Shows,TV-MA
1,Blood & Water,TV Show,TV Dramas,TV-MA
1,Blood & Water,TV Show,TV Mysteries,TV-MA
2,Ganglands,TV Show,Crime TV Shows,TV-MA
2,Ganglands,TV Show,International TV Shows,TV-MA
2,Ganglands,TV Show,TV Action & Adventure,TV-MA
3,Jailbirds New Orleans,TV Show,Docuseries,TV-MA
3,Jailbirds New Orleans,TV Show,Reality TV,TV-MA
4,Kota Factory,TV Show,International TV Shows,TV-MA



Top 10 géneros más frecuentes:


,count
genre,
International Movies,2752
Dramas,2427
Comedies,1674
International TV Shows,1351
Documentaries,869
Action & Adventure,859
TV Dramas,763
Independent Movies,756
Children & Family Movies,641



Clasificación de duración de películas:


,title,duration,duration_minutes,content_length,content_length_mask
0,Dick Johnson Is Dead,90 min,90.0,Contenido normal,Contenido normal
6,My Little Pony: A New Generation,91 min,91.0,Contenido normal,Contenido normal
7,Sankofa,125 min,125.0,Contenido largo,Contenido largo
9,The Starling,104 min,104.0,Contenido normal,Contenido normal
12,Je Suis Karl,127 min,127.0,Contenido largo,Contenido largo



Películas con más de 100 minutos, rating R y país United States:


,title,country,rating,duration_minutes
46,Safe House,"South Africa, United States, Japan",R,115.0
48,Training Day,United States,R,122.0
81,Kate,United States,R,106.0
122,In the Cut,"United Kingdom, Australia, France, United States",R,118.0
131,Blade Runner: The Final Cut,United States,R,117.0
133,Chappie,"South Africa, United States",R,121.0
135,Cliffhanger,"United States, Italy, France, Japan",R,113.0
136,Cold Mountain,"United States, Italy, Romania, United Kingdom",R,154.0
139,Do the Right Thing,United States,R,120.0
144,House Party,United States,R,104.0


,title,country,release_year,rating,duration_minutes
4253,Black Mirror: Bandersnatch,United States,2018,TV-MA,312 min
717,Headspace: Unwind Your Mind,nan,2021,TV-G,273 min
2491,The School of Mischief,Egypt,1973,TV-14,253 min
2487,No Longer kids,Egypt,1979,TV-14,237 min
2484,Lock Your Girls In,nan,1982,TV-PG,233 min
2488,Raya and Sakina,nan,1984,TV-14,230 min
166,Once Upon a Time in America,"Italy, United States",1984,R,229 min
7932,Sangam,India,1964,TV-14,228 min
1019,Lagaan,"India, United Kingdom",2001,PG,224 min
4573,Jodhaa Akbar,India,2008,TV-14,214 min


En esta sección se aplicaron distintas herramientas avanzadas de `pandas` para filtrar, transformar y reorganizar la información. Primero, se seleccionaron las películas agregadas después de 2018, lo que permite observar contenido incorporado más recientemente al catálogo.

También se filtraron títulos que contienen la palabra `love`, mostrando cómo trabajar con texto mediante `str.contains()`. Luego, se extrajo la duración numérica desde la columna `duration`, separando el valor cuantitativo del texto original. Este paso permite analizar la duración de las películas como una variable numérica.

Posteriormente, se utilizó `explode()` para separar los géneros listados en `listed_in`, generando una fila por cada género asociado a un título. Esto permite calcular de forma más clara cuáles son los géneros más frecuentes en Netflix. Además, se clasificaron las películas según su duración, identificando como contenido largo aquellas superiores a 120 minutos.

Finalmente, se aplicó un filtro más específico para identificar películas de más de 100 minutos, con rating `R` y producidas en Estados Unidos. El uso de `.style` permite presentar visualmente las películas más largas, facilitando la interpretación de los resultados.



### Pregunta Desafío

11. ¿Cuáles son las combinaciones más frecuentes de género y rating en el dataset?
    (Sugerencia: utilizar `value_counts` con `subset=["genre", "rating"]` después de aplicar `explode()`).



### Bonus: Análisis de duplicados y limpieza

12. ¿Existen películas con el mismo nombre (`title`) pero con distinto año de lanzamiento (`release_year`)?
13. ¿Cuántos títulos únicos hay en total en la columna `title`?





In [4]:
# Pregunta desafío y bonus

# 11. Combinaciones más frecuentes de género y rating
genre_rating_combinations = (
    df_genres
    .value_counts(subset=["genre", "rating"])
    .head(10)
    .reset_index(name="frequency")
)

print("Top 10 combinaciones más frecuentes de género y rating:")
display(genre_rating_combinations)

# 12. Películas con el mismo título, pero distinto año de lanzamiento
title_year_unique = df[["title", "release_year"]].drop_duplicates()

same_title_different_year = title_year_unique.loc[
    title_year_unique["title"].duplicated(keep=False)
].sort_values(["title", "release_year"])

print("\nTítulos repetidos con distinto año de lanzamiento:")
display(same_title_different_year)

print("\nCantidad de títulos con el mismo nombre y distinto año:")
print(same_title_different_year["title"].nunique())

# 13. Cantidad de títulos únicos
unique_titles = df["title"].nunique()

print("\nCantidad total de títulos únicos:")
print(unique_titles)

Top 10 combinaciones más frecuentes de género y rating:


,genre,rating,frequency
0,International Movies,TV-MA,1130
1,International Movies,TV-14,1065
2,Dramas,TV-MA,830
3,International TV Shows,TV-MA,714
4,Dramas,TV-14,693
5,International TV Shows,TV-14,472
6,Comedies,TV-14,465
7,TV Dramas,TV-MA,434
8,Comedies,TV-MA,431
9,Dramas,R,375



Títulos repetidos con distinto año de lanzamiento:


,title,release_year



Cantidad de títulos con el mismo nombre y distinto año:
0

Cantidad total de títulos únicos:
8807



Las combinaciones más frecuentes de género y rating permiten identificar qué tipos de contenido predominan en el catálogo según su clasificación etaria. Esta información puede ser útil para analizar la orientación del catálogo de Netflix hacia ciertos públicos y categorías de contenido.

El análisis de títulos repetidos permite detectar posibles remakes, contenidos con el mismo nombre o registros que podrían requerir revisión adicional. Para ello se compararon los títulos con sus años de lanzamiento, identificando aquellos casos en que un mismo nombre aparece asociado a más de un año.

Finalmente, la cantidad de títulos únicos entrega una medida más precisa del tamaño del catálogo, evitando contar repetidamente contenidos que puedan aparecer duplicados o con variaciones en alguna variable.

El análisis muestra que las combinaciones más frecuentes entre género y clasificación etaria se concentran principalmente en contenidos internacionales y clasificaciones para público adulto o adolescente. La combinación más repetida es **International Movies con clasificación TV-MA**, con **1130 títulos**, seguida por **International Movies con clasificación TV-14**, con **1065 títulos**. Esto indica que el catálogo de Netflix tiene una fuerte presencia de películas internacionales orientadas principalmente a audiencias adultas y adolescentes.

También destacan los **Dramas**, tanto con clasificación **TV-MA** como **TV-14**, lo que sugiere que este género tiene un peso importante dentro del catálogo. Además, aparecen categorías como **International TV Shows**, **Comedies** y **TV Dramas**, lo que evidencia una oferta diversa, pero concentrada en ciertos géneros de alta frecuencia.

Respecto al análisis de títulos repetidos con distinto año de lanzamiento, el resultado fue **0**. Esto significa que, dentro del dataset, no se encontraron títulos que tengan exactamente el mismo nombre pero distinto año de estreno. Por lo tanto, no se identifican posibles remakes, registros duplicados o títulos repetidos bajo distintos años de lanzamiento según el criterio utilizado.

Finalmente, la cantidad total de títulos únicos fue de **8807**. Este resultado indica que el catálogo analizado contiene 8807 títulos distintos. En términos generales, esto sugiere que la base de datos posee una alta variedad de contenido y que no existen duplicaciones relevantes en la columna `title`.

En conclusión, los resultados muestran que el catálogo de Netflix está fuertemente compuesto por contenido internacional, especialmente películas y series clasificadas para público adulto o adolescente. Además, la ausencia de títulos repetidos con distinto año refuerza la consistencia del dataset para este análisis.